### PR TITLE
CORDA-2595 - check that all attachments are trusted before loading

### PR DESCRIPTION
**Pull request by tudor-malene**
see: https://r3-cev.atlassian.net/browse/CORDA-2595

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt [hunk 1]
```diff
@@ -275,11 +275,15 @@ abstract class TransactionVerificationException(val txId: SecureHash, message: S
      * and because attachment classloaders are reused this is independent of any particular transaction.
      */
     @CordaSerializable
-    class PackageOwnershipException(txId: SecureHash, val attachmentHash: AttachmentId, val contractClass: String, val packageName: String) : TransactionVerificationException(txId,
-            """The Contract attachment JAR: $attachmentHash containing the contract: $contractClass is not signed by the owner of package $packageName specified in the network parameters.
+    class PackageOwnershipException(txId: SecureHash, val attachmentHash: AttachmentId, val invalidClassName: String, val packageName: String) : TransactionVerificationException(txId,
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt]: Note: This change is not really related to the current work. But it was an incorrect message due to the same assumption that led to this security issue.
(also reflected in the api file)

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 2]
```diff
@@ -107,13 +107,47 @@ class AttachmentsClassLoader(attachments: List<Attachment>,
...
+        for (attachment in attachments) {
+            if (!isZipOrJar(attachment)) throw TransactionVerificationException.InvalidAttachmentException(sampleTxId, attachment.id)
+        }
+
+        // Until we have a sandbox to run untrusted code we need to make sure that any loaded class file was whitelisted by the node administrator.
+        val untrusted = attachments
+                .filter(::containsClasses)
+                .filterNot { attachment ->
+                    when (attachment) {
+                        is ContractAttachment -> isUploaderTrusted(attachment.uploader)
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Why are we checking explicitly for `ContractAttachment` here?

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 3]
```diff
@@ -170,14 +172,14 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
         }
     }
 
-    private class FetchedAttachment(dataLoader: () -> ByteArray) : AbstractAttachment(dataLoader), SerializeAsToken {
+    private class FetchedAttachment(dataLoader: () -> ByteArray, uploader: String?) : AbstractAttachment(uploader, dataLoader), SerializeAsToken {
         override val id: SecureHash by lazy { attachmentData.sha256() }
 
-        private class Token(private val id: SecureHash) : SerializationToken {
-            override fun fromToken(context: SerializeAsTokenContext) = FetchedAttachment(context.attachmentDataLoader(id))
+        private class Token(private val id: SecureHash, private val uploader: String?) : SerializationToken {
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: why are we passing the `uploader` around to all these methods?

File: core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt [hunk 4]
```diff
@@ -115,7 +115,7 @@ class WireTransaction(componentGroups: List<ComponentGroup>, val privacySalt: Pr
     // Helper for deprecated toLedgerTransaction
     // TODO: revisit once Deterministic JVM code updated
     private val missingAttachment: Attachment by lazy {
-        object : AbstractAttachment({ byteArrayOf() }) {
+        object : AbstractAttachment("app", { byteArrayOf() } ) {
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt]: why a hardcoded value of "app" here?

Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt]: Should use constant DEPLOYED_CORDAPP_UPLOADER

File: core/src/test/kotlin/net/corda/core/internal/ResolveTransactionsFlowTest.kt [hunk 5]
```diff
@@ -155,7 +155,7 @@ class ResolveTransactionsFlowTest {
         }
         // TODO: this operation should not require an explicit transaction
         val id = megaCorpNode.transaction {
-            megaCorpNode.services.attachments.importAttachment(makeJar(), "test", null)
+            megaCorpNode.services.attachments.importAttachment(makeJar(), "TestDSL", null)
```
Comment (josecoll) [inline core/src/test/kotlin/net/corda/core/internal/ResolveTransactionsFlowTest.kt]: Should define and use constant from AbstractAttachment.

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt [hunk 6]
```diff
@@ -237,7 +237,7 @@ object DefaultKryoCustomizer {
 
                 return ContractAttachment.create(lazyAttachment, contract, additionalContracts, uploader, signers, version)
             } else {
-                val attachment = GeneratedAttachment(input.readBytesWithLength())
+                val attachment = GeneratedAttachment(input.readBytesWithLength(), "generated")
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt]: what is the implication of defining a new "generated" value here? 
(should always be using the constant set defined in AbstractAttachment)

File: core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt [hunk 7]
```diff
@@ -31,7 +31,7 @@ val TRUSTED_UPLOADERS = listOf(DEPLOYED_CORDAPP_UPLOADER, RPC_UPLOADER, "TestDSL
 fun isUploaderTrusted(uploader: String?): Boolean = uploader in TRUSTED_UPLOADERS
 
 @KeepForDJVM
-abstract class AbstractAttachment(dataLoader: () -> ByteArray) : Attachment {
+abstract class AbstractAttachment(val uploader: String?, dataLoader: () -> ByteArray) : Attachment {
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt]: `uploader` should be second parameter (in line with other definitions).

**Comment by tudor-malene** [comment]
> Can you explain why you are passing around the `uploader` field now (and setting lots of hardcoded default values in places)?

We are currently inferring trustability from the `uploader` of an attachment.

Currently, we have an unnecessary complex code structure around attachments, and this is not the time to refactor it. So the only way to bring the uploader field where it's required is to pass it around.

If you have a better suggestion, I'm happy to do something else.

File: node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt (hunk 6 above)
Comment (tudor-malene) [reply node/src/main/kotlin/net/corda/node/serialization/kryo/DefaultKryoCustomizer.kt]: afaict there is no implication. This class should go away with the refactoring.
The uploader is set correctly on the `ContractAttachment` that is created on top of this generated attachment thing.

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt (hunk 3 above)
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: constructors need them

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 2 above)
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: because of the attachment type hierarchy

**Comment by tudor-malene** [comment]
@mikehearn , @josecoll 
ready for re-review

**Comment by josecoll** [comment]
We need to add a QA test to exercise this new logic as its quite fundamental.
